Allow fx.Graph.owning_module to be used as attribute.

ydwu4 opened the pull request:
Summary:
The current behavior of owning_module setter is difficult to understand: it changes the owning_module to None if owners is not 0 but increments the owners count. If the owning_module is None, the owners count should be 0 as none of them is accessible. On the other hand, if the owners count increases, the owning_module should be a collection (e.g. a list).

This diff changes owning_module to be a normal attribute. The semantic is that graph can have **at most one** owning module and can be assigned to new module.

The alternative is to use a list to represent the owning_modules of a graph but it breaks backward compatibility and the exact use cases of having multiple owning_modules are not clear.

Test Plan: Test with CI.

Differential Revision: D40200624

zhxchen17 commented:
looks fine to me, the original setter behavior is very weird, will wait for all tests to pass.

tugsbayasgalan reviewed (commented):
LGTM if all tests pass

zhxchen17 commented:
/easycla

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot rebase

zhxchen17 commented:
@pytorchbot merge -f "test failure seems unrelated, and tests are flaky today."

Failed jobs
X linux-bionic-cuda11.6-py3.10-gcc7 / test (default, 3, 4, linux.4xlarge.nvidia.gpu)
Raw logs | more like this | Queued: 26m 11s | Duration: 24m 15s | Started: 12:25 PM
▶ RuntimeError: test_testing failed!
Classify failure: None
Broken Trunk
Test Flake
Broken Infra
Infra Flake
Network Error
Other
X linux-bionic-py3.7-clang9 / test (crossref, 1, 2, linux.2xlarge)
Raw logs | more like this | Queued: 6s | Duration: 28m 28s | Started: 12:14 PM
▶ RuntimeError: test_testing failed!
Classify failure: None
Broken Trunk
Test Flake
Broken Infra
Infra Flake
Network Error
Other
X linux-bionic-py3.7-clang9 / test (default, 2, 2, linux.2xlarge)
Raw logs | more like this | Queued: 5s | Duration: 34m 37s | Started: 12:14 PM
▶ AttributeError: module 'torch._C' has no attribute '_cuda_isInBadFork'
Classify failure: None
Broken Trunk
Test Flake
Broken Infra
Infra Flake
Network Error
Other
X linux-bionic-py3.7-clang9 / test (dynamo, 2, 2, linux.2xlarge)
Raw logs | more like this | Queued: 1s | Duration: 1h 1m 43s | Started: 12:14 PM
▶ AttributeError: module 'torch._C' has no attribute '_cuda_isInBadFork'
Classify failure: None
Broken Trunk
Test Flake
Broken Infra
Infra Flake
Network Error
Other
X linux-focal-py3.7-clang7-asan / test (default, 4, 5, linux.2xlarge)
Raw logs | more like this | Queued: 3s | Duration: 16m 56s | Started: 12:18 PM
▶ RuntimeError: test_testing failed!
Classify failure: None
Broken Trunk
Test Flake
Broken Infra
Infra Flake
Network Error
Other
X linux-focal-py3.7-gcc7 / test (default, 2, 2, linux.2xlarge)
Raw logs | more like this | Queued: 7s | Duration: 20m 33s | Started: 12:14 PM
▶ RuntimeError: test_testing failed!
Classify failure: None
Broken Trunk
Test Flake
Broken Infra
Infra Flake
Network Error
Other
X win-vs2019-cpu-py3 / test (default, 2, 2, windows.4xlarge)
Raw logs | more like this | Queued: 11s | Duration: 43m 44s | Started: 12:32 PM
▶ RuntimeError: test_testing failed!
Classify failure: None
Broken Trunk
Test Flake
Broken Infra
Infra Flake
Network Error
Other
X linux-bionic-cuda11.7-py3.10-gcc7 / test (default, 3, 4, linux.4xlarge.nvidia.gpu)
Raw logs | more like this | Queued: 24m 24s | Duration: 12m 43s | Started: 12:24 PM
▶ RuntimeError: test_testing failed!
Classify failure: None
Broken Trunk
Test Flake
Broken Infra
Infra Flake
Network Error
Other
X macos-12-py3-arm64 / test (default, 2, 2, macos-m1-12)
Raw logs | more like this | Queued: 2s | Duration: 12m 1s | Started: 12:36 PM
▶ RuntimeError: test_testing failed!
Classify failure: None
Broken Trunk
Test Flake
Broken Infra
Infra Flake
Network Error
Other
X macos-12-py3-x86-64 / test (default, 1, 2, macos-12)
Raw logs | more like this | Queued: 9s | Duration: 1h 14m 22s | Started: 12:42 PM
▶ AttributeError: module 'torch._C' has no attribute '_cuda_isInBadFork'
Classify failure: None
Broken Trunk
Test Flake
Broken Infra
Infra Flake
Network Error
Other
X parallelnative-linux-focal-py3.7-gcc7 / test (default, 2, 2, linux.2xlarge)
Raw logs | more like this | Queued: 7s | Duration: 18m 51s | Started: 12:14 PM
▶ RuntimeError: test_testing failed!
Classify failure:

ydwu4 committed:
Allow fx.Graph.owning_module to be used as an attribute. (#86822)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/86822

The current behavior of owning_module setter is difficult to understand: it changes the owning_module to None if owners is not 0 but increments the owners count. If the owning_module is None, the owners count should be 0 as none of them is accessible. On the other hand, if the owners count increases, the owning_module should be a collection (e.g. a list).

This diff changes owning_module to be a normal attribute. The semantic is that graph can have **at most one** owning module and can be assigned to new module.

The alternative is to use a list to represent the owning_modules of a graph but it breaks backward compatibility and the exact use cases of having multiple owning_modules are not clear.

Test Plan: Test with CI.

Reviewed By: tugsbayasgalan

Differential Revision: D40200624

fbshipit-source-id: d1f45df29625345692bdf36d268e16090db80163

ydwu4 commented:
@pytorchbot merge -f "test failure seems unrelated, and tests are flaky today."

2022-10-18T23:13:15.6938535Z Traceback (most recent call last):
2022-10-18T23:13:15.6938779Z   File "/opt/conda/lib/python3.7/site-packages/torch/_dynamo/variables/tensor.py", line 331, in create
2022-10-18T23:13:15.6938897Z     or proxy.node.target == torch._C._cuda_isInBadFork
2022-10-18T23:13:15.6939088Z AttributeError: module 'torch._C' has no attribute '_cuda_isInBadFork'
2022-10-18T23:13:15.6939094Z 
2022-10-18T23:13:15.6939162Z from user code:
2022-10-18T23:13:15.6939278Z    File "test_jit_fuser_te.py", line 520, in test_clamp_double
2022-10-18T23:13:15.6939550Z     s = self.checkScript(clamp_double, (x, eta), profiling=ProfilingMode.PROFILING, atol=1e-10, rtol=1e-5)
2022-10-18T23:13:15.6939555Z 
2022-10-18T23:13:15.6939683Z Set torch._dynamo.config.verbose=True for more information
2022-10-18T23:13:15.6939746Z ==========
2022-10-18T23:13:15.6939799Z ok (0.152s)
2022-10-18T23:13:15.6940057Z   test_clamp_int (__main__.TestTEFuserDynamic) ... WON'T CONVERT test_clamp_int test_jit_fuser_te.py line 523 
2022-10-18T23:13:15.6940119Z due to: 
2022-10-18T23:13:15.6940209Z Traceback (most recent call last):
2022-10-18T23:13:15.6940453Z   File "/opt/conda/lib/python3.7/site-packages/torch/_dynamo/variables/tensor.py", line 331, in create
2022-10-18T23:13:15.6940569Z     or proxy.node.target == torch._C._cuda_isInBadFork
2022-10-18T23:13:15.6940758Z AttributeError: module 'torch._C' has no attribute '_cuda_isInBadFork'
2022-10-18T23:13:15.6940763Z 
2022-10-18T23:13:15.6940822Z from user code:
2022-10-18T23:13:15.6940935Z    File "test_jit_fuser_te.py", line 530, in test_clamp_int
2022-10-18T23:13:15.6941094Z     s = self.checkScript(clamp_int, (x, eta), profiling=ProfilingMode.PROFILING)
2022-10-18T23:13:15.6941099Z 
2022-10-18T23:13:15.6941228Z Set torch._dynamo.config.verbose=True for more information
2022-10-18T23:13:15.6941291Z ==========
2022-10-18T23:13:15.6941357Z ok (0.143s)
2022-10-18T23:13:15.6941488Z   test_comparison_eq_ne (__main__.TestTEFuserDynamic) ... ok (0.309s)
2022-10-18T23:13:15.6941617Z   test_comparison_ge_le (__main__.TestTEFuserDynamic) ... ok (0.310s)
2022-10-18T23:13:15.6941731Z   test_comparison_gt_lt (__main__.TestTEFuserDynamic) ... ok (0.308s)
2022-10-18T23:13:15.6941878Z   test_concat (__main__.TestTEFuserDynamic) ... ok (0.576s)
2022-10-18T23:13:15.6942007Z   test_concat_invariant (__main__.TestTEFuserDynamic) ... ok (0.776s)
2022-10-18T23:13:15.6942180Z   test_constant_chunk_shapes (__main__.TestTEFuserDynamic) ... skip: TODO: chunk dynamic shapes (0.024s)
2022-10-18T23:13:15.6942441Z   test_conv2d (__main__.TestTEFuserDynamic) ... skip: don't run conv with dynamic shapes (0.006s)
2022-10-18T23:13:15.6942695Z   test_conv2d_depthwise (__main__.TestTEFuserDynamic) ... skip: don't run conv with dynamic shapes (0.006s)
2022-10-18T23:13:15.6942842Z   test_cuda_half (__main__.TestTEFuserDynamic) ... skip: fuser requires CUDA (0.001s)
2022-10-18T23:13:15.6943080Z   test_dims (__main__.TestTEFuserDynamic) ... WON'T CONVERT test_dims test_jit_fuser_te.py line 2027 
2022-10-18T23:13:15.6943133Z due to: 
2022-10-18T23:13:15.6943223Z Traceback (most recent call last):
2022-10-18T23:13:15.6943472Z   File "/opt/conda/lib/python3.7/site-packages/torch/_dynamo/variables/tensor.py", line 331, in create
2022-10-18T23:13:15.6943592Z     or proxy.node.target == torch._C._cuda_isInBadFork
2022-10-18T23:13:15.6943783Z AttributeError: module 'torch._C' has no attribute '_cuda_isInBadFork'
2022-10-18T23:13:15.6943788Z 
2022-10-18T23:13:15.6943857Z from user code:
2022-10-18T23:13:15.6943966Z    File "test_jit_fuser_te.py", line 2032, in test_dims
2022-10-18T23:13:15.6944075Z     script = self.checkScript(eager, (x, y))
2022-10-18T23:13:15.6944080Z 
2022-10-18T23:13:15.6944195Z Set torch._dynamo.config.verbose=True for more information
2022-10-18T23:13:15.6944256Z ==========
2022-10-18T23:13:15.6944319Z ok (0.141s)
2022-10-18T23:13:15.6944575Z   test_disabled (__main__.TestTEFuserDynamic) ... WON'T CONVERT test_disabled test_jit_fuser_te.py line 1163 
2022-10-18T23:13:15.6944637Z due to: 
2022-10-18T23:13:15.6944723Z Traceback (most recent call last):
2022-10-18T23:13:15.6944971Z   File "/opt/conda/lib/python3.7/site-packages/torch/_dynamo/variables/tensor.py", line 331, in create
2022-10-18T23:13:15.6945078Z     or proxy.node.target == torch._C._cuda_isInBadFork
2022-10-18T23:13:15.6945270Z AttributeError: module 'torch._C' has no attribute '_cuda_isInBadFork'

ydwu4 commented:
@pytorchmergebot merge -f "test failure seems unrelated, and tests are flaky today."

2022-10-18T23:13:15.6938535Z Traceback (most recent call last):
2022-10-18T23:13:15.6938779Z File "/opt/conda/lib/python3.7/site-packages/torch/_dynamo/variables/tensor.py", line 331, in create
2022-10-18T23:13:15.6938897Z or proxy.node.target == torch._C._cuda_isInBadFork
2022-10-18T23:13:15.6939088Z AttributeError: module 'torch._C' has no attribute '_cuda_isInBadFork'
2022-10-18T23:13:15.6939094Z
2022-10-18T23:13:15.6939162Z from user code:
2022-10-18T23:13:15.6939278Z File "test_jit_fuser_te.py", line 520, in test_clamp_double
2022-10-18T23:13:15.6939550Z s = self.checkScript(clamp_double, (x, eta), profiling=ProfilingMode.PROFILING, atol=1e-10, rtol=1e-5)
2022-10-18T23:13:15.6939555Z
2022-10-18T23:13:15.6939683Z Set torch._dynamo.config.verbose=True for more information
2022-10-18T23:13:15.6939746Z ==========
2022-10-18T23:13:15.6939799Z ok (0.152s)
2022-10-18T23:13:15.6940057Z test_clamp_int (main.TestTEFuserDynamic) ... WON'T CONVERT test_clamp_int test_jit_fuser_te.py line 523
2022-10-18T23:13:15.6940119Z due to:
2022-10-18T23:13:15.6940209Z Traceback (most recent call last):
2022-10-18T23:13:15.6940453Z File "/opt/conda/lib/python3.7/site-packages/torch/_dynamo/variables/tensor.py", line 331, in create
2022-10-18T23:13:15.6940569Z or proxy.node.target == torch._C._cuda_isInBadFork
2022-10-18T23:13:15.6940758Z AttributeError: module 'torch._C' has no attribute '_cuda_isInBadFork'
2022-10-18T23:13:15.6940763Z
2022-10-18T23:13:15.6940822Z from user code:
2022-10-18T23:13:15.6940935Z File "test_jit_fuser_te.py", line 530, in test_clamp_int
2022-10-18T23:13:15.6941094Z s = self.checkScript(clamp_int, (x, eta), profiling=ProfilingMode.PROFILING)
2022-10-18T23:13:15.6941099Z
2022-10-18T23:13:15.6941228Z Set torch._dynamo.config.verbose=True for more information
2022-10-18T23:13:15.6941291Z ==========

ydwu4 commented:
@pytorchmergebot merge -f "test failure seems unrelated, and tests are flaky today."

Try again with updated permission.
